xscen/diagnostics.py

```diff
@@ -2,110 +2,290 @@
+    if start_date is not None:
+        # Create cf_time objects to compare the dates
+        start_date = xr.cftime_range(
+            start=start_date, periods=1, freq="D", calendar=ds.time.dt.calendar
+        )[0]
+        if not ((ds_start <= start_date) and (ds_end > start_date)):
+            err = f"The start date is not at least {start_date}. Received {ds.time.min().values.astype('datetime64[m]')}."
+            if "start_date" in raise_on:
+                raise ValueError(err)
+            else:
+                warnings.warn(err, UserWarning, stacklevel=1)
+    if end_date is not None:
+        # Create cf_time objects to compare the dates
+        end_date = xr.cftime_range(
+            start=end_date, periods=1, freq="D", calendar=ds.time.dt.calendar
+        )[0]
+
+        if not ((ds_start < end_date) and (ds_end >= end_date)):
+            err = f"The end date is not at least {end_date}. Received {ds.time.max().values.astype('datetime64[m]')}."
+            if "end_date" in raise_on:
+                raise ValueError(err)
+            else:
+                warnings.warn(err, UserWarning, stacklevel=1)
+
+    # Check variables
+    if variables_and_units is not None:
+        for v in variables_and_units:
+            if v not in ds:
+                raise ValueError(f"The variable '{v}' is missing.")
+            if ds[v].attrs.get("units", None) != variables_and_units[v]:
+                xc.core.units.check_units(
+                    ds[v], variables_and_units[v]
+                )  # Will always raise an error if the units are not compatible
+                err = f"The variable '{v}' does not have the expected units '{variables_and_units[v]}'. Received '{ds[v].attrs['units']}'."
+                if "variables_and_units" in raise_on:
+                    raise ValueError(err)
+                else:
+                    warnings.warn(err, UserWarning, stacklevel=1)
+
+    # Check CF conventions
+    if cfchecks is not None:
+        cfchecks = deepcopy(cfchecks)
+        for v in cfchecks:
+            for check in cfchecks[v]:
+                if check == "check_valid":
+                    cfchecks[v][check]["var"] = ds[v]
+                elif check == "cfcheck_from_name":
+                    cfchecks[v][check].setdefault("varname", v)
+                    cfchecks[v][check]["vardata"] = ds[v]
+                else:
+                    raise ValueError(f"Check '{check}' is not in xclim.")
+                if "cfchecks" in raise_on:
+                    warnings.simplefilter("error")
+                    try:
+                        getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
+                    except UserWarning as e:
+                        raise ValueError(e)
+                    warnings.resetwarnings()
+                else:
+                    getattr(xc.core.cfchecks, check)(**cfchecks[v][check])
+
+    inferred_freq = xr.infer_freq(ds.time)
+    if freq is not None:
```

Comment:
```suggestion
    if freq is not None:
        inferred_freq = xr.infer_freq(ds.time)
```
In case the dataset has no "time" coordinate ? Or that the time coord has only 1 or 2 elements.

Reply:
This was because `inferred_freq` was reused for the `missing` checks, but your suggestions work too.